### PR TITLE
Add UL2 data sampling and pretraining

### PR DESCRIPTION
**Pull request by janEbert**
This adds pretraining using [UL2](https://arxiv.org/abs/2205.05131) for both encoder-decoder, non-causal decoder-only, and causal decoder-only models.
I have not yet run large-scale tests to see if it yields the desired training improvements, but I wanted to give others the option to take a look at the code already.

### CODE REVIEW

**Comment by janEbert** [comment]
Previously, I truncated sequences so the maximum amount of duplicated `extra_id` tokens would fit in and still be accepted by the model, losing some data all the time. I now changed it so the program just errors out and asks the user to put in a longer sequence length for the model.

This is probably a worse/undesired solution, so I kept the other code in for now (but commented).

File: megatron/data/dataset_utils.py [hunk 1]
```diff
@@ -235,18 +251,24 @@ def create_masked_lm_predictions(tokens,
...
-    if not geometric_dist:
+    if sampling_style is SamplingStyle.POISSON:
         # Note(mingdachen):
         # By default, we set the probilities to favor shorter ngram sequences.
         pvals = 1. / np.arange(1, max_ngrams + 1)
         pvals /= pvals.sum(keepdims=True)
         if favor_longer_ngram:
             pvals = pvals[::-1]
+    elif sampling_style is SamplingStyle.NORMAL:
+        normal_mean = (max_ngrams + 1) / 2
```
Comment (Muennighoff) [inline megatron/data/dataset_utils.py:262]: `normal_mean` is not used it seems

Comment (janEbert) [reply megatron/data/dataset_utils.py:262]: It's used here. :)
https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/358/commits/d8db18922e97e7b8bf1e45182665bc73f223bfc1#diff-e1d14be32f4489a01cb8d571804fbba003f7f90715ef3cb3a27d9099e0245d6fR298

**Comment by janEbert** [comment]
There were several issues still remaining in the UL2 implementation, most notably that I only tested for micro batch sizes of 1, which when increased made the decoder-only models fail. :p
Also most notably in terms of the UL2 sampling, there was an issue regarding the S-denoisers, in which the mean was not correctly positioned, leading to shorter masks than desired.

The implementation also more closely follows the `seqio` implementation in the UL2 paper now, which omits the single `extra_id` token for the Prefix-LM task, which we previously added.

**Comment by janEbert** [comment]
I can finally report results... Comparing standard T5 training vs training with UL2 or UL2R, results in lm-eval-harness were almost always better with UL2/UL2R. Which should mean this code does improve evaluation results. :)
